src/spam-protection/spam-protection-action-creator.ts

```diff
@@ -0,0 +1,88 @@
+    constructor(
+        private _googleRecaptcha: GoogleRecaptcha,
+        private _requestSender: SpamProtectionRequestSender
+    ) {}
+
+    initialize(options?: SpamProtectionOptions): ThunkAction<SpamProtectionAction, InternalCheckoutSelectors> {
+        return store => Observable.create((observer: Observer<SpamProtectionAction>) => {
+            const spamProtectionElementId = options ? options.containerId : 'spamProtectionContainer';
+
+            if (!options && !document.getElementById(spamProtectionElementId)) {
+                const spamProtectionElement = document.createElement('div');
+                spamProtectionElement.setAttribute('id', spamProtectionElementId);
+                document.body.append(spamProtectionElement);
+            }
+
+            const state = store.getState();
+            const config = state.config.getConfig();
+
+            if (!config) {
+                throw new MissingDataError(MissingDataErrorType.MissingCheckoutConfig);
+            }
+
+            observer.next(createAction(SpamProtectionActionType.InitializeRequested, undefined));
+
+            const recaptchaSitekey = config.storeConfig.checkoutSettings.googleRecaptchaSitekey;
+
+            return this._googleRecaptcha.load(spamProtectionElementId, recaptchaSitekey)
+                .then(() => {
+                    observer.next(createAction(SpamProtectionActionType.InitializeSucceeded));
+                    observer.complete();
+                })
+                .catch(error => {
+                    observer.error(createErrorAction(SpamProtectionActionType.InitializeFailed, error, spamProtectionElementId));
+                });
+        });
+    }
+
+    validate(): ThunkAction<SpamProtectionAction, InternalCheckoutSelectors> {
+        return store => {
+            const state = store.getState();
+            const storeConfig = state.config.getStoreConfig();
+
+            if (!storeConfig) {
+                throw new MissingDataError(MissingDataErrorType.MissingCheckoutConfig);
+            }
+
+            const { isSpamProtectionEnabled } = storeConfig.checkoutSettings;
+
+            if (!isSpamProtectionEnabled) {
+                return empty();
+            }
+
+            return concat(
+                of(createAction(SpamProtectionActionType.ExecuteRequested, undefined)),
+                this._googleRecaptcha.execute()
+                    .pipe(take(1))
+                    .pipe(switchMap(({ error, token }) => {
+                        if (error || !token) {
+                            return throwError(createErrorAction(SpamProtectionActionType.SubmitFailed, error));
+                        }
+
+                        const checkout = state.checkout.getCheckout();
+
+                        if (!checkout) {
+                            throw new MissingDataError(MissingDataErrorType.MissingCheckout);
+                        }
+
+                        this._requestSender.validate(checkout.id, token);
```

Comment:
This looks like an asynchronous request. Should we wait for it to complete?

Reply:
Oops, I missed this.